### DISCUSSION TITLE
Archiving workflows

### DISCUSSION THREAD

**Discussion by maahutch** [redivis/meta#192] [Feature requests]
Hi everyone, 

We've been discussing how we can archive work on Redivis that uses proprietary data. Since a lot of our data is licensed we can't just take a copy of the entire dataset and add it to our digital archive. However, this limits the reproducibility of research given we aren't able to store the data our researchers are using for the long-term.

I my previous position, I had contributed to a platform called [Cadre](https://cadre.iu.edu/resources) that was trying to be a web data storage and analytics platform, similar to Redivis but with an additional focus in reproducibility. Our platform gave users the option to archive subsets of data generated by queries, along with any analytical code into a docker container which we then archived and assigned a DOI. Then, if the archived container weren't used for a significant period of time, the data would be replaced by the query used to extract it from the larger dataset. 

I was wondering if it might be possible to introduce something like this to the Redivis platform? Whether it would be possible to package metadata about a dataset, query code used to generate a subset and analytic code used in research? That way we could archive the packaged code in our digital archive and make it discoverable to other researchers. 

I appreciate this is a pretty significant change to the Redivis platform but I wanted to put the idea out there to see if there was any interest in something like this or if Redivis is already working on a way to make the platform's workflows reproducible.

Thanks,

**Comment by imathews** [comment]
Hi @maahutch,

Thanks so much for opening this. This is definitely something our team has thought quite a bit about, and we definitely see research reproducibility as core tenet of Redivis. I think the best way to respond might be to break out what we support now, what we envision in the near future, and then some final items that might merit further discussion.

### Today
Perhaps most importantly, all datasets on Redivis are versioned, with an eye towards storing these version diffs as efficiently (aka cheaply) as possible. This allows for data owners to continuously update their data, while ensuring that analyses run on previous versions of the data continue to produce the same results. 

Moreover, Redivis "projects" allow for researchers to capture, persist, and export their entire data workflow in a reproducible manner. Any data transformations are codified in SQL, while notebooks can define their specific dependencies and be exported in common formats. Redivis does automatically archive derivative tables within projects after a period of disuse, but that table can be quickly unarchived since all the code used to create the output table is persisted.

One caveat to this reproducibility story is that data owners do have the option to delete datasets (or specific versions of a dataset). There are really two reasons this might happen: 1) legal or other contractual obligations; and 2) to reduce storage costs. When data deletion occurs, Redivis immediately marks any derivative tables in researchers' projects as archived, meaning that they can view table metadata, as well as any code / transformations used to produce those tables, but can not view, export, or query the actual data. So in this case, the researcher still has all of their code, but in order to reproduce or build on their analysis, they would have to find the dataset from another source that hasn't been deleted.

### On the roadmap
I think the biggest piece missing today is version control within researchers' projects on Redivis. While projects do offer a permanent, reproducible data workflow, there's generally not a way for me to see my project as it was at a previous point in time. This feature is currently on our roadmap, and something that I hope we can make significant progress on this year. 

Versioned projects open up a lot of exciting possibilities, beyond the obvious utility of researchers being able to jump back to a previous state in their workflow. For example, we plan to offer the ability to mint a DOI for a particular version of a project, support adding projects to other projects, and even think about how new datasets could be born from the output of a previous project.

### For further discussion
You bring up an interesting question re: packaging data alongside final analyses, such that deleting the source dataset wouldn't affect the ability to rerun the final analysis. This inevitably brings up the question of storage and legal constraints — somebody ultimately needs to pay for these storage costs, and there are certainly circumstances where it is necessary to expunge all derivatives of a dataset when a license expires, etc. I think the cost question is generally manageable (either outputs are small enough so as to be negligible, or the institution / researcher could foot the bill), but the contractual rules around data ownership might be a bit trickier.

I do want to think more about how we can thread this needle. For example, we've talked internally about a world where researchers could request that a subset of their outputs be "submitted for publication" — that is, the data owner would acknowledge that these outputs are appropriately aggregated / redacted so as to not leak any information from the source dataset(s). Once approved, these outputs would now belong to the researcher — even if the dataset gets deleted, or the researcher otherwise loses access. This would allow the researcher to package the output data, figures, and all code, and presumably offer this package alongside their final publication.

-------

As always, would love to hear your thoughts on any of the above, and happy to dig into any of these points in finer detail!
